doc/guides/using-internal-errors.md

```diff
@@ -0,0 +1,112 @@
+# Using the internal/errors.js Module
+
+## What is internal/errors.js
+
+The `require('internal/errors')` module is an internal-only module that can be
+used to produce `Error`, `TypeError` and `RangeError` instances that use a
+static, permanent error code and an optionally parameterized message. The
+intent of the module is to make errors thrown by Node.js core to be more
+consistent and to allow changes to error messages to be made without requiring
+a semver-major label.
+
+Switching an existing error to use the `internal/errors` module must be
+considered a semver-major change. However, once using `internal/errors`,
+changes to `internal/errors` error messages will be handled as semver-minor
+or patch.
+
+## Using internal/errors.js
+
+The `internal/errors` module exposes three custom `Error` classes as a
+`message` function.
+
+### Class: errors.Error(key[, args...])
+
+* `key` {String} The static error identifier
+* `args... {Any} Zero or more optional arguments
```

Comment:
Look at the `message()` function. When a key is registered using the `E()` function, it's associated value is either a string with optional util.format tokens or a function that returns a string. Inside `message()`, the key lookup returns this value. If the value is a string, it is passed through `util.format()`, otherwise the arguments are passed to the function, allowing more customized formatting for the error.

Again, this is *not* a user facing API. This is for use only *within* core and would be used anywhere we throw an internal error.